### ISSUE TITLE
Routing table entry and security groups keeps being rebuilt

### ISSUE THREAD

**Issue by FransUrbo** [hashicorp/terraform#5423]
I'm having problem with Terraform modifying existing routes and security groups every time it runs. This also leads to the routes, _some_ of them (!!), not working and some do. Running **apply** again will recreate all of them (or modify them in such a way that they don't work any more), but this time others will work and others not (have not looked to see if there's a pattern).

I have five VPCs: **OneA** (10.128/16), **OneB** (10.129/16), **Two** (10.130/16), **Three** (10.131/16) and **Four** (10.132/16). And more coming...

VPC **OneA** peers with everyone, VPC **OneB** peers with **Two** and **Three**. **Four** doesn't peer with anyone (it's supposed to be a self-contained entity).
# Peering config on **OneA**:

```
########################
#   Peering (on OneA)  #
########################

# ---- OneB
resource "aws_vpc_peering_connection" "vpc_peer_OneA-OneB" {
  peer_owner_id = "${aws_security_group.OneA.owner_id}"
  peer_vpc_id = "${var.OneB_vpc_id}"
  vpc_id = "${aws_vpc.OneA.id}"
  auto_accept = true

  tags {
    Name = "OneA-OneB"
  }
}
resource "aws_route" "peering_OneA-OneB" {
  route_table_id = "${aws_route_table.public.id}"
  destination_cidr_block = "${var.OneB_vpc_cidr}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneA-OneB.id}"
}
resource "aws_route" "peering_OneB-OneA" {
  route_table_id = "${var.OneB_vpc_rtable}"
  destination_cidr_block = "${aws_vpc.OneA.cidr_block}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneA-OneB.id}"
}

# ---- Two
resource "aws_vpc_peering_connection" "vpc_peer_OneA-Two" {
  peer_owner_id = "${aws_security_group.OneA.owner_id}"
  peer_vpc_id = "${var.Two_vpc_id}"
  vpc_id = "${aws_vpc.OneA.id}"
  auto_accept = true

  tags {
    Name = "OneA-Two"
  }
}
resource "aws_route" "peering_OneA-Two" {
  route_table_id = "${aws_route_table.public.id}"
  destination_cidr_block = "${var.Two_vpc_cidr}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneA-Two.id}"
}
resource "aws_route" "peering_Two-OneA" {
  route_table_id = "${var.Two_vpc_rtable}"
  destination_cidr_block = "${aws_vpc.OneA.cidr_block}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneA-Two.id}"
}

# ---- Three
resource "aws_vpc_peering_connection" "vpc_peer_OneA-Three" {
  peer_owner_id = "${aws_security_group.OneA.owner_id}"
  peer_vpc_id = "${var.Three_vpc_id}"
  vpc_id = "${aws_vpc.OneA.id}"
  auto_accept = true

  tags {
    Name = "OneA-Three"
  }
}
resource "aws_route" "peering_OneA-Three" {
  route_table_id = "${aws_route_table.public.id}"
  destination_cidr_block = "${var.Three_vpc_cidr}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneA-Three.id}"
}
resource "aws_route" "peering_Three-OneA" {
  route_table_id = "${var.Three_vpc_rtable}"
  destination_cidr_block = "${aws_vpc.OneA.cidr_block}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneA-Three.id}"
}
```
# Peering config on **OneB**:

```
########################
#   Peering (on OneB)  #
########################

# ---- Two
resource "aws_vpc_peering_connection" "vpc_peer_OneB-Two" {
  peer_owner_id = "${var.OneA_owner_id}"
  peer_vpc_id = "${var.Two_vpc_id}"
  vpc_id = "${aws_vpc.OneB.id}"
  auto_accept = true

  tags {
    Name = "OneB-Two"
  }
}
resource "aws_route" "peering_OneB-Two" {
  route_table_id = "${aws_route_table.public.id}"
  destination_cidr_block = "${var.Two_vpc_cidr}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneB-Two.id}"
}
resource "aws_route" "peering_Two-OneB" {
  route_table_id = "${var.Two_vpc_rtable}"
  destination_cidr_block = "${aws_vpc.OneB.cidr_block}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneB-Two.id}"
}

# ---- Three
resource "aws_vpc_peering_connection" "vpc_peer_OneB-Three" {
  peer_owner_id = "${var.OneA_owner_id}"
  peer_vpc_id = "${var.Three_vpc_id}"
  vpc_id = "${aws_vpc.OneB.id}"
  auto_accept = true

  tags {
    Name = "OneB-Three"
  }
}
resource "aws_route" "peering_OneB-Three" {
  route_table_id = "${aws_route_table.public.id}"
  destination_cidr_block = "${var.Three_vpc_cidr}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneB-Three.id}"
}
resource "aws_route" "peering_Three-OneB" {
  route_table_id = "${var.Three_vpc_rtable}"
  destination_cidr_block = "${aws_vpc.OneB.cidr_block}"
  vpc_peering_connection_id = "${aws_vpc_peering_connection.vpc_peer_OneB-Three.id}"
}
```
# Security Group on **OneA**:

```
resource "aws_security_group" "OneA" {
  name = "vpc-OneA"
  description = "Default security group that allows inbound and outbound traffic from all instances in the VPC"
  vpc_id = "${aws_vpc.OneA.id}"

  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    self        = true
  }

  /* All communication from the OneB VPC */
  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.129.0.0/16"]
  }

  /* All communication from the Two VPC */
  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.130.0.0/16"]
  }

  /* All communication from the Three VPC */
  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.131.0.0/16"]
  }

  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    self        = true
  }

  /* All communication from the OneB VPC */
  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.129.0.0/16"]
  }

  /* All communication from the Two VPC */
  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.130.0.0/16"]
  }

  /* All communication from the Three VPC */
  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.131.0.0/16"]
  }

  tags {
    Name = "OneA-vpc"
  }
}
```
# Security Group on **OneB**:

```
resource "aws_security_group" "OneB" {
  name = "vpc-OneB"
  description = "Default security group that allows inbound and outbound traffic from all instances in the VPC"
  vpc_id = "${aws_vpc.OneB.id}"

  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    self        = true
  }

  /* All communication from the OneA VPC */
  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.128.0.0/16"]
  }

  /* All communication from the Two VPC */
  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.130.0.0/16"]
  }

  /* All communication from the Three VPC */
  ingress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.131.0.0/16"]
  }

  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    self        = true
  }

  /* All communication to the OneA VPC */
  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.128.0.0/16"]
  }

  /* All communication to the Two VPC */
  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.130.0.0/16"]
  }

  /* All communication to the Three VPC */
  egress {
    from_port   = "0"
    to_port     = "0"
    protocol    = "-1"
    cidr_blocks = ["10.131.0.0/16"]
  }

  tags {
    Name = "OneB-vpc"
  }
}
```

If I taint all of these and run <code>apply</code> I get:

```
* InvalidParameterValue: cannot disassociate the main route table association rtbassoc-40f44324
        status code: 400, request id:
* InvalidParameterValue: cannot disassociate the main route table association rtbassoc-20fd4b44
        status code: 400, request id:
* InvalidParameterValue: cannot disassociate the main route table association rtbassoc-41f44325
        status code: 400, request id:
* InvalidParameterValue: cannot disassociate the main route table association rtbassoc-7bf4431f
        status code: 400, request id:
* InvalidParameterValue: cannot disassociate the main route table association rtbassoc-b6e355d2
        status code: 400, request id:
* timeout while waiting for state to become '[success]'
* timeout while waiting for state to become '[success]'
* timeout while waiting for state to become '[success]'
* timeout while waiting for state to become '[success]'
* timeout while waiting for state to become '[success]'
```

I have not any other "aws_route" resources configured in my Terraform setup. Only a simple <code>aws_main_route_table_association</code> in each VPC/module (where <code>module_name</code> is the same as the VPC name):

```
resource "aws_main_route_table_association" "<module_name>" {
  vpc_id = "${aws_vpc.<module_name>.id}"
  route_table_id = "${aws_route_table.public.id}"
}
```
# First apply

After having Terraform getting out of sync with reality (another issue) and doing a <code>destroy</code> and then running <code>apply</code> from scratch, hosts in both VPCs **Two** and **Three** could trace/ping a host in VPC **OneA** and everything is dandy ("just fine" :)!
# Second apply

Running it again, neither hosts could trace a **OneA** host.

See http://bayour.com/misc/Terraform-Recreates-Routes-And-SecurityGroups-1.txt for the output for the _second_ <code>apply</code>.
# Third apply

Running it a third time gives: http://bayour.com/misc/Terraform-Recreates-Routes-And-SecurityGroups-2.txt and this time, everything works again.

The only thing I see as obvious (and I might not see everything :), this time, <code>module.OneA.aws_route_table.public</code> is gone...

---

Looking at the AWS Console/Peering Connections (when routing works - i.e. after the first apply) I see that all peering connections have the corresponding route tables (and points to the correct entries etc). The security groups all look a-ok.

Running <code>apply</code> again (second apply), _ALL_ route table entries for _ALL_ VPCs are removed!

**Comment by FransUrbo** [comment]
I've been trying to reproduce this twice now (from scratch), and both times Terraform have also recreated three of the four instances in **OneB**.

# Second apply
http://bayour.com/misc/Terraform-Recreates-Routes-And-SecurityGroups-3.txt

* Routing does NOT work

# Third apply
http://bayour.com/misc/Terraform-Recreates-Routes-And-SecurityGroups-4.txt

* Routing DOES work (again)

**Comment by jen20** [comment]
Hi @FransUrbo! Thanks for opening this issue and providing a detailed description. Could you confirm which version of Terraform you are running during these tests? Thanks!

**Comment by FransUrbo** [comment]
v0.6.11

**Comment by FransUrbo** [comment]
*bump* Anything about this yet? It's really annoying :)

Btw, I now run latest git (```Terraform v0.6.15-dev (b20f680cbf0505825e6ae173c5986243683d8eb3+CHANGES)```).

**Comment by augusto** [comment]
Hi Guys! Has anyone found a solution to this? I just bumped into this problem with v0.8.1. The symptoms I see are:

- 1st apply: TF adds an entry on the routing table (for VPC peering) and a NAT entry.
- 2nd apply: TF removes the VPC peering and NAT  entries from the routing table (apps start failing)
- 3rd apply: TF re-adds the VPC and NAT entries.
- rinse and repeat.

**Comment by jen20** [comment]
It looks like this is closed in the same manner as #7303.
